scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,175 @@
+
+This process could be either manual or automated.
+
+### Building versioned bundles
+
+#### Online mode
+
+The GitHub workflow will have a map with "bundle version" as keys and "source reference" as values, similar to Table 1 above.  This map can be stored statically in the matrix definition or as a file in the source repository.
+
+A source reference can be one of the following:
+
+1.  A branch reference, `heads/<branch-name>`, example: `heads/release-4.6`
+2.  A tag reference, `tags/<tag-name>`, example: `tags/4.6.0`.
+3.  A StackRox release, `release/<stackrox-release>`, example: `release/4.6.0`
+
+Both (1.) and (2.) will determine a specific commit to use to build the vulnerability exporter.
+
+(3.) will select a desired release build based on a StackRox target release.  The workflow will find the best matching Git tag for the specified `<stackrox-release>`, by checking if the version exists as a tag or matching it against release candidate (`-rc`) or patch (`.x`) tags.
+
+For each version in the mapping, a bundle will be built based on the determined git ref and stored at `https://definitions.stackrox.io/v4/<bundle-version>/vulnerabilities.zip`.
+
+In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release reference (e.g., `release/4.6.0`).  Notice that the key is still updated if the version is the same.   If it's new, it will be added, and the previous version will stay.
+
+For example, in release cut 4.6 we would look into `scanner/VULNERABILITY_VERSION`, let's say its contents are:
+
+```
+$ cat canner/VULNERABILITY_VERSION
+v1
+```
+
+The mapping should be updated so `v1` points to the current release branch reference.  Let's say the mapping was:
+
+bundle version | git reference
+--- | ---
+`dev` | `master`
+`v1` | `release/4.5.0`
+
+Then, `v1` is updated to the new release (in this example, `4.6.0`):
+
+bundle version | git reference
+--- | ---
+`dev` | `master`
+`v1` | `release/4.6.0`
+
+If in 4.6 the bundle version was updated to:
+
+```
+$ cat canner/VULNERABILITY_VERSION
+v2
+```
+
+On 4.6.0 release cut, the final mapping would be:
+
+bundle version | git reference
+--- | ---
+`dev` | `master`
+`v1` | `release/4.5.0`
+`v2` | `release/4.6.0`
+
+#### Offline mode
+
+Per-release offline bundles will continue to be published.  They will be versioned by StackRox releases, but they will not be based on Y-stream (`4.5`) but on Z-stream patch releases (`4.5.0`).  The bundle will contain the correct vulnerability bundle for that particular Z-stream release.  The `manifest.json` file will continue to have a "version" attribute pointing out the Y-stream for that offline bundle, but two other fields will be added: "vulnerability_version" and "release_versions".  For example, for patch release `4.6.1` the offline bundle could be:
+
+```
+{
+  "version": "4.6",
+  "created": "2024-08-20T21:00:30+00:00",
+  "release_versions": ["4.6.0", "4.6.1", ...],
```

Comment:
the `...` implies there may be more. Why would 4.6.1 know about more than just 4.6.0 and 4.6.1?

Also, can you explain why we introduce this? How does it differ from what we did before with just `version`?

Reply:
> the ... implies there may be more. Why would 4.6.1 know about more than just 4.6.0 and 4.6.1?

Not sure I follow your question. I think it might be just a confusion in the example: By adding `...` I am trying to say that the list will go until the latest release.

This will allow Central to validate if this bundle supports its version. If Central is running `4.5.0` in this example, it will reject the bundle.

Original "version" shows the bundle schema version now for versioned bundles. It's not useful for Central to decide to reject or accept the bundle because Central doesn't have visibility into the Scanner's schema version. And that's by design.

Reply:
>> the ... implies there may be more. Why would 4.6.1 know about more than just 4.6.0 and 4.6.1?
>
>Not sure I follow your question. I think it might be just a confusion in the example: By adding ... I am trying to say that the list will go until the latest release.

Yeah, I was confused because the previous paragraph said this was an example for `4.6.1`, so I did not expect to see anything more than `4.6.0` and `4.6.1`. Can you perhaps update the patch version for the example and then make this concrete?

> I considered getting rid of vulnerability_version and use only version. In 4.[45].x bundles it will be "4.4" or "4.5", in schema versioned bundles, it will be be the schema version (e.g. "v1"), FYI.

I don't recall what this file looked like previously. Was there a `vulnerability_version` field which you ended up removing?

Reply:
Yes, I ended up removing the `vulnerability_version` based on the consideration you quoted. Essentially, the "version" field is the Y-stream version pre 4.6; moving forward, it's the vulnerability bundle schema version (like "v1", "v2", so on so forth...).